sensorprocess/sensorprocess_test.go

```diff
@@ -0,0 +1,412 @@
+				timeout:          timeout,
+				sensorName:       sensorName,
+				currentReading:   currentReading,
+				readingTimestamp: readingTimestamp,
+			}
+			calls = append(calls, args)
+			if len(calls) == 1 {
+				return errUnknown
+			}
+			if len(calls) < 4 {
+				return cartofacade.ErrUnableToAcquireLock
+			}
+			return nil
+		}
+		addSensorReadingFromReplaySensor(cancelCtx, reading, readingTimestamp, config)
+		test.That(t, len(calls), test.ShouldEqual, 4)
+		for i, args := range calls {
+			t.Logf("addSensorReadingArgsHistory %d", i)
+			test.That(t, args.timeout, test.ShouldEqual, config.Timeout)
+			test.That(t, args.sensorName, test.ShouldEqual, config.LidarName)
+			test.That(t, args.currentReading, test.ShouldResemble, reading)
+			test.That(t, args.readingTimestamp, test.ShouldResemble, readingTimestamp)
+		}
+		cancelFunc()
+	})
+}
+
+func TestAddSensorReadingLiveReadings(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+	cf := cartofacade.Mock{}
+	reading := []byte("12345")
+	readingTimestamp := time.Now().UTC()
+	// When addSensorReading blocks for longer than the data rate
+	config := Config{
+		Logger:      logger,
+		Cartofacade: &cf,
+		LidarName:   "good_lidar",
+		DataRateMs:  200,
+		Timeout:     10 * time.Second,
+	}
+
+	t.Run("When AddSensorReading blocks for more than the DataRateMs and succeeds, returns 0", func(t *testing.T) {
+		cf.AddSensorReadingFunc = func(
+			ctx context.Context,
+			timeout time.Duration,
+			sensorName string,
+			currentReading []byte,
+			readingTimestamp time.Time,
+		) error {
+			time.Sleep(1 * time.Second)
+			return nil
+		}
+
+		timeToSleep := addSensorReadingFromLiveReadings(context.Background(), reading, readingTimestamp, config)
+		test.That(t, timeToSleep, test.ShouldEqual, 0)
+	})
+
+	t.Run("AddSensorReading slower than DataRateMs and returns lock error, returns 0", func(t *testing.T) {
+		cf.AddSensorReadingFunc = func(
+			ctx context.Context,
+			timeout time.Duration,
+			sensorName string,
+			currentReading []byte,
+			readingTimestamp time.Time,
+		) error {
+			time.Sleep(1 * time.Second)
+			return cartofacade.ErrUnableToAcquireLock
+		}
+
+		timeToSleep := addSensorReadingFromLiveReadings(context.Background(), reading, readingTimestamp, config)
+		test.That(t, timeToSleep, test.ShouldEqual, 0)
+	})
+
+	t.Run("When AddSensorReading blocks for more than the DataRateMs and returns an unexpected error, returns 0", func(t *testing.T) {
+		cf.AddSensorReadingFunc = func(
+			ctx context.Context,
+			timeout time.Duration,
+			sensorName string,
+			currentReading []byte,
+			readingTimestamp time.Time,
+		) error {
+			time.Sleep(1 * time.Second)
+			return errUnknown
+		}
+
+		timeToSleep := addSensorReadingFromLiveReadings(context.Background(), reading, readingTimestamp, config)
+		test.That(t, timeToSleep, test.ShouldEqual, 0)
+	})
+
+	t.Run("AddSensorReading faster than the DataRateMs and succeeds, returns int <= DataRateMs", func(t *testing.T) {
```

Comment:
"returns int <= DataRateMs" -> "time to sleep is <= DataRateMs"?

Reply:
done